First Chunking Example

In [2]:
text = """
Python is a programming language.
Django is a web framework.
FastAPI is used for APIs.
Machine learning uses data.
"""

chunks = text.split(".")
for item in chunks:
    if item.strip():  # This skips the final empty item
        print(item.strip())

print(chunks)

Python is a programming language
Django is a web framework
FastAPI is used for APIs
Machine learning uses data
['\nPython is a programming language', '\nDjango is a web framework', '\nFastAPI is used for APIs', '\nMachine learning uses data', '\n']


In [3]:
text = """
Python is a programming language.
Django is a web framework.
FastAPI is used for APIs.
Machine learning uses data.
"""

chunks = text.split(".")

for i, chunk in enumerate(chunks):
    print(f"Chunk {i}: {chunk}")

Chunk 0: 
Python is a programming language
Chunk 1: 
Django is a web framework
Chunk 2: 
FastAPI is used for APIs
Chunk 3: 
Machine learning uses data
Chunk 4: 



In [4]:
chunks = [chunk.strip() for chunk in text.split(".") if chunk.strip()]
print(f"{chunks}")

['Python is a programming language', 'Django is a web framework', 'FastAPI is used for APIs', 'Machine learning uses data']


chunk.strip()


Removes:

- spaces

- tabs

- newlines (\n)

PDF

↓

Extract Text

↓

Chunking

↓

View Chunks

↓

Understand why chunking is needed

In [5]:
from google.colab import files

uploaded = files.upload()

Saving NBayesLogReg.pdf to NBayesLogReg.pdf


In [6]:
!pip install -q pypdf langchain

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 346.6/346.6 kB 22.1 MB/s eta 0:00:00


In [7]:
from pypdf import PdfReader

reader = PdfReader("NBayesLogReg.pdf")

print("Total Pages:", len(reader.pages))

text = ""

for page in reader.pages:
    text += page.extract_text()

print("\nTotal Characters:", len(text))

Total Pages: 18

Total Characters: 36668


Explanation

from pypdf import PdfReader
- Imports the PDF reader library.

reader = PdfReader("NBayesLogReg.pdf")
- Loads the PDF file into memory.

print("Total Pages:", len(reader.pages))
- Counts the total number of pages in the PDF.

text = ""
- Creates an empty string to store extracted text.

for page in reader.pages:
- Loops through every page in the PDF.

text += page.extract_text()
- Extracts text from the current page and appends it to the text variable.

print("\nTotal Characters:", len(text))
- Displays the total number of characters extracted from the PDF.

In [8]:
!pip install -q langchain-text-splitters

In [9]:
!pip list | grep langchain

langchain                                1.3.6
langchain-core                           1.4.3
langchain-protocol                       0.0.16
langchain-text-splitters                 1.1.2


In [10]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

print("Import successful")

Import successful


In [11]:
splitter = RecursiveCharacterTextSplitter(
    chunk_size=500,
    chunk_overlap=100
)

chunks = splitter.split_text(text)

print("Total Chunks:", len(chunks))

Total Chunks: 95


Explanation

splitter = RecursiveCharacterTextSplitter(...)
- Creates a text splitter object.

chunk_size=500
- Maximum size of each chunk is 500 characters.

chunk_overlap=100
- Repeats 100 characters from the previous chunk.

chunks = splitter.split_text(text)
- Splits the PDF text into chunks.

print("Total Chunks:", len(chunks))
- Displays the total number of chunks created.

In [12]:
for i, chunk in enumerate(chunks[:3]):
    print(f"\n----- Chunk {i+1} -----\n")
    print(chunk)


----- Chunk 1 -----

CHAPTER 3
GENERATIVE AND DISCRIMINATIVE
CLASSIFIERS:
NAIVE BAYES AND LOGISTIC REGRESSION
Machine Learning
Copyright c⃝2017. Tom M. Mitchell. All rights reserved.
*DRAFT OF October 1, 2020*
*PLEASE DO NOT DISTRIBUTE WITHOUT AUTHOR’S PERMISSION*
This is a rough draft chapter intended for inclusion in the upcoming sec-
ond edition of the textbook Machine Learning, T.M. Mitchell, McGraw Hill.
You are welcome to use this for educational purposes, but do not dupli-

----- Chunk 2 -----

You are welcome to use this for educational purposes, but do not dupli-
cate or repost it on the internet. To distribute this to others, simply point
them to http://www.cs.cmu.edu/∼tom/mlbook/NBayesLogReg.pdf. For online
copies of this and other materials related to this book, visit the web site
www.cs.cmu.edu/∼tom/mlbook.html.
Please send suggestions for improvements, or suggested exercises, to
Tom.Mitchell@cmu.edu.
1 Learning Classifiers based on Bayes Rule

----- Chunk 3 -----

Tom.Mit

Explanation

chunks[:3]
- Selects the first 3 chunks.

enumerate(...)
- Provides chunk number and content.

print(chunk)
- Displays the chunk text.

36,668 characters

↓

RecursiveCharacterTextSplitter

↓

95 chunks

↓

Ready for Embeddings

In [13]:
for i in chunks[0:5]:
  print(i+"\n")

CHAPTER 3
GENERATIVE AND DISCRIMINATIVE
CLASSIFIERS:
NAIVE BAYES AND LOGISTIC REGRESSION
Machine Learning
Copyright c⃝2017. Tom M. Mitchell. All rights reserved.
*DRAFT OF October 1, 2020*
*PLEASE DO NOT DISTRIBUTE WITHOUT AUTHOR’S PERMISSION*
This is a rough draft chapter intended for inclusion in the upcoming sec-
ond edition of the textbook Machine Learning, T.M. Mitchell, McGraw Hill.
You are welcome to use this for educational purposes, but do not dupli-

You are welcome to use this for educational purposes, but do not dupli-
cate or repost it on the internet. To distribute this to others, simply point
them to http://www.cs.cmu.edu/∼tom/mlbook/NBayesLogReg.pdf. For online
copies of this and other materials related to this book, visit the web site
www.cs.cmu.edu/∼tom/mlbook.html.
Please send suggestions for improvements, or suggested exercises, to
Tom.Mitchell@cmu.edu.
1 Learning Classifiers based on Bayes Rule

Tom.Mitchell@cmu.edu.
1 Learning Classifiers based on Bayes Rule
Here we